### PR TITLE
perf: improve efficiency on processing instructions

### LINKED ISSUE #2560: improve efficiency on processing instructions
#### Problem
To gather `ComputeBudgetLimits` for a transaction, it currently iterates through all instructions to find its program_id, then compare to compute_budget::id(). This is inefficient. 

#### Proposed Solution
- dedup instruction.program_id_index to avoid lookup for same program_id multiple times,
- filter non-builtin instructions early,

### PR DESCRIPTION
**Pull request by tao-stones**
#### Problem
#2560 

#### Summary of Changes
1. add filter to avoid duplicated process of same program_id_index;
2. add static filter to filter out non-builtin programs early;

Fixes #2560

### CODE REVIEW

**Comment by tao-stones** [comment]
bench results below. Because the repeated work so far is `compute_budget::check_id(program_id)` for every instructions, which is just a cheap compareson, the perf improvement for `4 compute-budget instructions` and `4 dummy builtins` are not noticeable. (it would be when pubkey hashing is involved in next PR).

But still, the static MAYBE_BUILTIN_KEY filter improved `4 dummy Instructions`, while `0 instructions` regressed due to overhead without any gain.

<details>
<summary>bench on master before PR</summary>

```
     Running benches/process_compute_budget_instructions.rs (target/release/deps/process_compute_budget_instructions-1b953003a907ddf6)
bench_process_compute_budget_instructions_empty/0 instructions
                        time:   [5.4891 µs 5.4930 µs 5.4975 µs]
                        thrpt:  [186.27 Melem/s 186.42 Melem/s 186.55 Melem/s]
                 change:
                        time:   [-0.1118% +0.0838% +0.2649%] (p = 0.39 > 0.05)
                        thrpt:  [-0.2642% -0.0838% +0.1119%]
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

bench_process_compute_budget_instructions_no_builtins/4 dummy Instructions
                        time:   [16.933 µs 16.944 µs 16.957 µs]
                        thrpt:  [60.390 Melem/s 60.434 Melem/s 60.475 Melem/s]
                 change:
                        time:   [-0.2684% -0.0836% +0.0871%] (p = 0.37 > 0.05)
                        thrpt:  [-0.0870% +0.0837% +0.2691%]
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  1 (1.00%) low mild
  6 (6.00%) high mild
  6 (6.00%) high severe

bench_process_compute_budget_instructions_compute_budgets/4 compute-budget instructions
                        time:   [25.644 µs 25.665 µs 25.691 µs]
                        thrpt:  [39.858 Melem/s 39.899 Melem/s 39.931 Melem/s]
                 change:
                        time:   [-0.5227% -0.2684% -0.0223%] (p = 0.04 < 0.05)
                        thrpt:  [+0.0223% +0.2691% +0.5254%]
                        Change within noise threshold.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe

bench_process_compute_budget_instructions_builtins/4 dummy builtins
                        time:   [17.202 µs 17.211 µs 17.221 µs]
                        thrpt:  [59.462 Melem/s 59.496 Melem/s 59.527 Melem/s]
                 change:
                        time:   [-0.3604% -0.1489% +0.0329%] (p = 0.15 > 0.05)
                        thrpt:  [-0.0329% +0.1491% +0.3617%]
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  2 (2.00%) low mild
  5 (5.00%) high mild
  7 (7.00%) high severe

Benchmarking bench_process_compute_budget_instructions_mixed/355 mixed instructions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.6s, enable flat sampling, or reduce sample count to 60.
bench_process_compute_budget_instructions_mixed/355 mixed instructions
                        time:   [1.1012 ms 1.1024 ms 1.1043 ms]
                        thrpt:  [927.27 Kelem/s 928.88 Kelem/s 929.90 Kelem/s]
                 change:
                        time:   [-0.3482% -0.2090% -0.0521%] (p = 0.00 < 0.05)
                        thrpt:  [+0.0521% +0.2095% +0.3494%]
                        Change within noise threshold.
Found 10 outliers among 100 measurements (10.00%)
  5 (5.00%) high mild
  5 (5.00%) high severe
```

</details>

<details>
<summary>bench on PR</summary>

```
     Running benches/process_compute_budget_instructions.rs (target/release/deps/process_compute_budget_instructions-a2fc49a61149f446)
bench_process_compute_budget_instructions_empty/0 instructions
                        time:   [8.9378 µs 8.9599 µs 8.9820 µs]
                        thrpt:  [114.01 Melem/s 114.29 Melem/s 114.57 Melem/s]
                 change:
                        time:   [+62.462% +62.817% +63.128%] (p = 0.00 < 0.05)
                        thrpt:  [-38.698% -38.581% -38.447%]
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild

bench_process_compute_budget_instructions_no_builtins/4 dummy Instructions
                        time:   [14.316 µs 14.327 µs 14.342 µs]
                        thrpt:  [71.401 Melem/s 71.472 Melem/s 71.527 Melem/s]
                 change:
                        time:   [-15.454% -15.335% -15.216%] (p = 0.00 < 0.05)
                        thrpt:  [+17.947% +18.112% +18.278%]
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

bench_process_compute_budget_instructions_compute_budgets/4 compute-budget instructions
                        time:   [26.105 µs 26.153 µs 26.211 µs]
                        thrpt:  [39.067 Melem/s 39.154 Melem/s 39.226 Melem/s]
                 change:
                        time:   [+1.6651% +1.8963% +2.1342%] (p = 0.00 < 0.05)
                        thrpt:  [-2.0896% -1.8610% -1.6378%]
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  4 (4.00%) high severe

bench_process_compute_budget_instructions_builtins/4 dummy builtins
                        time:   [17.145 µs 17.163 µs 17.182 µs]
                        thrpt:  [59.597 Melem/s 59.665 Melem/s 59.727 Melem/s]
                 change:
                        time:   [-0.5084% -0.3666% -0.2216%] (p = 0.00 < 0.05)
                        thrpt:  [+0.2221% +0.3679% +0.5110%]
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild

bench_process_compute_budget_instructions_mixed/355 mixed instructions
                        time:   [497.61 µs 498.21 µs 498.82 µs]
                        thrpt:  [2.0528 Melem/s 2.0554 Melem/s 2.0579 Melem/s]
                 change:
                        time:   [-55.020% -54.944% -54.873%] (p = 0.00 < 0.05)
                        thrpt:  [+121.60% +121.95% +122.32%]
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild
```

</details>

File: runtime-transaction/src/compute_budget_program_id_filter.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+use {solana_builtins_default_costs::MAYBE_BUILTIN_KEY, solana_sdk::pubkey::Pubkey};
+
+pub(crate) struct ComputeBudgetProgramIdFilter {
+    // array of 256 slots for all possible program_id_index (as u8),
+    // each slot indicates if a program_id_index has not been checked (eg, None),
+    // or already checked with result (eg, Some(result)) that can be reused.
+    flags: [Option<bool>; 256],
```
Comment (apfitzge) [inline runtime-transaction/src/compute_budget_program_id_filter.rs]: See comment at: https://github.com/anza-xyz/agave/blob/master/transaction-view/src/static_account_keys_meta.rs#L13

`program_id_index` must be static account keys. Due to `PACKET_DATA_SIZE` constraint, we can only have up to 38 static account keys.

Based on that, I think we can make this array smaller.

Comment (apfitzge) [reply runtime-transaction/src/compute_budget_program_id_filter.rs]: this assumes that the instructions we are inspecting have already been sanitized.

**Review by apfitzge — DISMISSED** [review]
Changes look fine to me.
Made a comment about reducing overhead if we can assume instruction sanitization (maybe a quick rename of fn?)

File: runtime-transaction/src/compute_budget_program_id_filter.rs (hunk 1 above)
Comment (tao-stones) [reply runtime-transaction/src/compute_budget_program_id_filter.rs]: Awesome! 19a8f724d8be5127b0d955ba17f6ddd85d629b65 sets array size to MAX_STATIC_ACCOUNTS_PER_PACKET, also expects sanitized program_id_index.

<details>
<summary>bench improves after allocating smaller vec</summary>

```
Running benches/process_compute_budget_instructions.rs (target/release/deps/process_compute_budget_instructions-5f6e3aaf6ce9e51b)
bench_process_compute_budget_instructions_empty/0 instructions
                        time:   [6.1550 µs 6.1608 µs 6.1671 µs]
                        thrpt:  [166.04 Melem/s 166.21 Melem/s 166.37 Melem/s]
                 change:
                        time:   [-31.187% -31.006% -30.792%] (p = 0.00 < 0.05)
                        thrpt:  [+44.492% +44.939% +45.322%]
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) low mild
  2 (2.00%) high mild
  3 (3.00%) high severe

bench_process_compute_budget_instructions_no_builtins/4 dummy Instructions
                        time:   [11.730 µs 11.735 µs 11.740 µs]
                        thrpt:  [87.222 Melem/s 87.262 Melem/s 87.299 Melem/s]
                 change:
                        time:   [-18.297% -18.172% -18.045%] (p = 0.00 < 0.05)
                        thrpt:  [+22.017% +22.208% +22.394%]
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) low mild
  3 (3.00%) high mild
  6 (6.00%) high severe

bench_process_compute_budget_instructions_compute_budgets/4 compute-budget instructions
                        time:   [24.310 µs 24.320 µs 24.330 µs]
                        thrpt:  [42.087 Melem/s 42.106 Melem/s 42.123 Melem/s]
                 change:
                        time:   [-7.2499% -7.0161% -6.7625%] (p = 0.00 < 0.05)
                        thrpt:  [+7.2530% +7.5455% +7.8166%]
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  5 (5.00%) high mild
  7 (7.00%) high severe

bench_process_compute_budget_instructions_builtins/4 dummy builtins
                        time:   [16.051 µs 16.069 µs 16.089 µs]
                        thrpt:  [63.645 Melem/s 63.725 Melem/s 63.798 Melem/s]
                 change:
                        time:   [-6.5258% -6.3697% -6.1985%] (p = 0.00 < 0.05)
                        thrpt:  [+6.6081% +6.8031% +6.9814%]
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  6 (6.00%) high mild
  8 (8.00%) high severe

bench_process_compute_budget_instructions_mixed/355 mixed instructions
                        time:   [491.68 µs 506.11 µs 525.54 µs]
                        thrpt:  [1.9485 Melem/s 2.0233 Melem/s 2.0826 Melem/s]
                 change:
                        time:   [-0.9431% +0.9639% +2.9959%] (p = 0.47 > 0.05)
                        thrpt:  [-2.9087% -0.9547% +0.9521%]
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) high mild
  7 (7.00%) high severe
```

</details>

File: builtins-default-costs/src/lib.rs [hunk 2]
```diff
@@ -40,3 +40,17 @@ lazy_static! {
     .cloned()
     .collect();
 }
+
+lazy_static! {
+    /// A table of 256 booleans indicates whether the first `u8` of a Pubkey exists in
+    /// BUILTIN_INSTRUCTION_COSTS. If the value is true, the Pubkey might be a builtin key;
+    /// if false, it cannot be a builtin key. This table allows for quick filtering of
+    /// builtin program IDs without the need for hashing.
+    pub static ref MAYBE_BUILTIN_KEY: [bool; 256] = {
```
Comment (apfitzge) [inline builtins-default-costs/src/lib.rs:49]: curious if you benched without this pre-filter?

File: transaction-view/src/static_account_keys_meta.rs [hunk 3]
```diff
@@ -10,7 +10,7 @@ use {
 // This means the maximum number of 32 byte keys is 38.
 // 38 as an min-sized encoded u16 is 1 byte.
 // We can simply read this byte, if it's >38 we can return None.
-const MAX_STATIC_ACCOUNTS_PER_PACKET: u8 =
+pub const MAX_STATIC_ACCOUNTS_PER_PACKET: u8 =
```
Comment (apfitzge) [inline transaction-view/src/static_account_keys_meta.rs:13]: can't comment below, if this module is now `pub` - probably want the `StaticAccountKeysMeta` struct to be `pub(crate)` instead.

Comment (tao-stones) [reply transaction-view/src/static_account_keys_meta.rs:13]: b07ddccc51c67613ffa965e9c6e23f3882124a0e

File: builtins-default-costs/src/lib.rs (hunk 2 above)
Comment (tao-stones) [reply builtins-default-costs/src/lib.rs:49]: If I disable checking this at `check_program_id()`, then bench again, all tests see regression except `4 dummy builtins`, as expected.

<details>
<summary>run without MAYBE_BUILTIN_KEY check </summary>

```
    Running benches/process_compute_budget_instructions.rs (target/release/deps/process_compute_budget_instructions-5f6e3aaf6ce9e51b)
bench_process_compute_budget_instructions_empty/0 instructions
                        time:   [6.6563 µs 6.6645 µs 6.6728 µs]
                        thrpt:  [153.46 Melem/s 153.65 Melem/s 153.84 Melem/s]
                 change:
                        time:   [+10.493% +10.766% +11.062%] (p = 0.00 < 0.05)
                        thrpt:  [-9.9606% -9.7192% -9.4968%]
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

bench_process_compute_budget_instructions_no_builtins/4 dummy Instructions
                        time:   [12.297 µs 12.301 µs 12.306 µs]
                        thrpt:  [83.208 Melem/s 83.244 Melem/s 83.274 Melem/s]
                 change:
                        time:   [+2.8059% +3.0991% +3.4057%] (p = 0.00 < 0.05)
                        thrpt:  [-3.2935% -3.0059% -2.7293%]
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  8 (8.00%) high severe

bench_process_compute_budget_instructions_compute_budgets/4 compute-budget instructions
                        time:   [25.557 µs 25.580 µs 25.603 µs]
                        thrpt:  [39.995 Melem/s 40.031 Melem/s 40.067 Melem/s]
                 change:
                        time:   [+3.7466% +4.0747% +4.3736%] (p = 0.00 < 0.05)
                        thrpt:  [-4.1903% -3.9152% -3.6113%]
                        Performance has regressed.
Found 15 outliers among 100 measurements (15.00%)
  3 (3.00%) low severe
  1 (1.00%) low mild
  5 (5.00%) high mild
  6 (6.00%) high severe

bench_process_compute_budget_instructions_builtins/4 dummy builtins
                        time:   [13.438 µs 13.473 µs 13.523 µs]
                        thrpt:  [75.721 Melem/s 76.003 Melem/s 76.200 Melem/s]
                 change:
                        time:   [-16.351% -16.176% -15.960%] (p = 0.00 < 0.05)
                        thrpt:  [+18.991% +19.297% +19.547%]
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) high mild
  8 (8.00%) high severe

bench_process_compute_budget_instructions_mixed/355 mixed instructions
                        time:   [521.62 µs 521.94 µs 522.32 µs]
                        thrpt:  [1.9605 Melem/s 1.9619 Melem/s 1.9631 Melem/s]
                 change:
                        time:   [+6.3906% +7.1169% +7.9015%] (p = 0.00 < 0.05)
                        thrpt:  [-7.3229% -6.6440% -6.0067%]
                        Performance has regressed.
Found 15 outliers among 100 measurements (15.00%)
  3 (3.00%) high mild
  12 (12.00%) high severe

```

</details>

**Review by apfitzge — APPROVED** [review]

**Pull request merged**